### PR TITLE
Authentication mechanism for external applications

### PR DESCRIPTION
**Pull request by JulianKniephoff**
See commit message(s) and the added docs for more details. Commit-by-commit-review isn't really necessary since there are only two and the second one is tiny. But it's still possible. :shrug:

### CODE REVIEW

File: backend/src/auth/mod.rs [hunk 1]
```diff
@@ -326,6 +367,24 @@ impl HasRoles for User {
     }
 }
 
+impl HasRoles for AuthContext {
+    fn roles(&self) -> &[String] {
+        static TRUSTED_ROLES: Lazy<[String; 1]> = Lazy::new(|| [ROLE_ANONYMOUS.into()]);
```
Comment (LukasKalbertodt) [inline backend/src/auth/mod.rs:372]: Trusted external tools only have the `ROLE_ANONYMOUS`? My intuition would be that they are rather powerful? Can you explain this again?

File: backend/src/auth/mod.rs [hunk 2]
```diff
@@ -326,6 +367,24 @@ impl HasRoles for User {
...
+        &*TRUSTED_ROLES
+    }
+
+    fn is_moderator(&self, auth_config: &AuthConfig) -> bool {
+        // TODO: This is only implemented because the `mount`-API
+        // currently uses other APIs that require moderator privileges.
+        use AuthContext::*;
```
Comment (LukasKalbertodt) [inline backend/src/auth/mod.rs:379]: See above regarding `use`

File: docs/auth/README.md [hunk 3]
```diff
@@ -126,3 +126,20 @@ Tobira's "login" buttons in the header will then directly link to that URL.
...
 Of course, then you define how a login attempt looks like and what to do on a successful login.
 
+
+## Authenticating Opencast against Tobira
+
+If you happen to use the integration of Tobira into the Opencast Admin-UI
+to directly mount newly created series, Opencast has to authenticate
+against Tobira as well. This does not use most of the mechanisms above.
+Instead, Tobira and Opencast have to share a secret, which Opencast
+sends to Tobira under the `x-tobira-trusted_key`-header.
```
Comment (LukasKalbertodt) [inline docs/auth/README.md:136]:
```suggestion
sends to Tobira under the `x-tobira-trusted-key`-header.
```

File: backend/src/api/context.rs [hunk 4]
```diff
@@ -13,7 +13,7 @@ use crate::{
 /// The context that is accessible to every resolver in our API.
 pub(crate) struct Context {
     pub(crate) db: Transaction,
-    pub(crate) user: Option<User>,
+    pub(crate) authn: AuthContext,
```
Comment (LukasKalbertodt) [inline backend/src/api/context.rs:16]: Why `authn`? Like I know "authz" and "authn" stand for authorization and authentification respectively, so that would be authentification... but. Why not just `auth` or `auth_context`? I find it a bit weird to explicitly exclude authorization here? The field is used for both.

File: backend/src/api/mutation.rs [hunk 5]
```diff
@@ -182,6 +183,10 @@ impl Mutation {
         // because doing it like this duplicates some work
         // like checking moderator rights, input validity, etc.
 
+        if !matches!(context.authn, AuthContext::TrustedExternal) {
```
Comment (LukasKalbertodt) [inline backend/src/api/mutation.rs:186]: In this case I wouldn't use `matches!` but rather derive `PartialEq, Eq` for `AuthContext` (and likely `User`). Then you can just write `if context.authn != AuthContext::TrustedExternal`

File: backend/src/api/query.rs [hunk 6]
```diff
@@ -70,15 +70,18 @@ impl Query {
 
     /// Returns the current user.
     fn current_user(context: &Context) -> Option<&User> {
-        context.user.as_ref()
+        match context.authn {
+            AuthContext::User(ref user) => Some(user),
+            _ => None,
+        }
```
Comment (LukasKalbertodt) [inline backend/src/api/query.rs:76]: As already mentioned, since [RFC 2005](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html), `ref` is almost never needed anymore and it's more idiomatic to instead do this:

```suggestion
        match &context.authn {
            AuthContext::User(user) => Some(user),
            _ => None,
        }
```

File: backend/src/auth/mod.rs [hunk 7]
```diff
@@ -135,10 +152,34 @@ pub(crate) struct User {
...
+        db: &Client,
+    ) -> Result<Self, PgError> {
+
+        if let Some(request_trusted_key) = headers.get("x-tobira-trusted-key") {
+            if let Some(ref config_trusted_key) = auth_config.trusted_key {
+                if config_trusted_key.expose_secret() == request_trusted_key {
+                    return Ok(Self::TrustedExternal);
+                }
+            }
+        }
```
Comment (LukasKalbertodt) [inline backend/src/auth/mod.rs:177]: I've thought a bit about this as it seems one can write this much shorter. And one kind of can with `map_or` and `zip` or `matches!`. But in practice, I feel none of the other versions improve upon this. So I think it's fine just leaving it as is. I think only one version I would also consider fine:

```rust
        if auth_config.trusted_key.is_some() {
            if auth_config.trusted_key == headers.get("x-tobira-trusted-key") {
                return Ok(Self::TrustedExternal);        
            }
        }
```

But if you leave the current solution, remove `ref` (see other comment about this):

```suggestion
        if let Some(request_trusted_key) = headers.get("x-tobira-trusted-key") {
            if let Some(config_trusted_key) = &auth_config.trusted_key {
                if config_trusted_key.expose_secret() == request_trusted_key {
                    return Ok(Self::TrustedExternal);
                }
            }
        }
```

File: backend/src/auth/mod.rs [hunk 8]
```diff
@@ -93,6 +94,14 @@ pub(crate) struct AuthConfig {
     #[config(default = "30d", deserialize_with = crate::config::deserialize_duration)]
     pub(crate) session_duration: Duration,
 
+    /// A shared secret for **trusted** external applications.
+    /// Send this value as the `x-tobira-trusted_key`-header
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply backend/src/auth/mod.rs:98]: How about `trusted-external-key` or `external-trusted-key`? I think putting `tool` in there is a bit wordy and kinda sounds unnatural. :thinking:

File: backend/src/auth/mod.rs [hunk 9]
```diff
@@ -135,10 +152,34 @@ pub(crate) struct User {
 }
 
 /// Returns a representation of the optional username useful for logging.
-pub(crate) fn debug_log_username(session: &Option<User>) -> Cow<'static, str> {
+pub(crate) fn debug_log_username(session: &AuthContext) -> Cow<'static, str> {
+    use AuthContext::*;
     match session {
-        None => "none".into(),
-        Some(user) => format!("'{}'", user.username).into(),
+        Anonymous => "none".into(),
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply backend/src/auth/mod.rs:158]: Yip, copy&paste error.

File: backend/src/auth/mod.rs [hunk 10]
```diff
@@ -135,10 +152,34 @@ pub(crate) struct User {
 }
 
 /// Returns a representation of the optional username useful for logging.
-pub(crate) fn debug_log_username(session: &Option<User>) -> Cow<'static, str> {
+pub(crate) fn debug_log_username(session: &AuthContext) -> Cow<'static, str> {
+    use AuthContext::*;
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply backend/src/auth/mod.rs:156]: I can always be wrong, but I looked at lots of different rust code, so let's say yes, plz change ^_^

File: backend/src/auth/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply backend/src/auth/mod.rs:158]: Good question. Can't really tell you I think. I see that it could be confusing. I saw `None` being defined in others' peoples code as well. So :shrug: Maybe let's go with anonymous, then it's not confusing at least and "anonymous" is descriptive as well.

[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply backend/src/auth/mod.rs:158]: Let's keep `Anonymous` for now. I will fix the string.

File: backend/src/api/context.rs (hunk 4 above)
Comment (JulianKniephoff) [reply backend/src/api/context.rs:16]: True, why not `auth`. :shrug:

File: backend/src/auth/mod.rs (hunk 7 above)
Comment (JulianKniephoff) [reply backend/src/auth/mod.rs:177]: Your suggestion doesn't work as is, because it tries to compare an `Option<Secret<String>>` with a an `Option<&HeaderValue>`, so you would have to `unwrap` or `map` the first while you already know that there is a `Some` in there. This is always kind of a red flag to me. So I would say let's keep it? I'll fix the `ref` thing, though.
